sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/ModelHelper.java

```diff
@@ -466,6 +485,171 @@ public static long getBlobLength(BlobDownloadHeaders headers) {
             : ChunkedDownloadUtils.extractTotalBlobLength(headers.getContentRange());
     }
 
+    public static boolean checkBlobDoesNotExistStatusCode(Throwable t) {
+        if (t instanceof BlobStorageException) {
+            BlobStorageException s = (BlobStorageException) t;
+            return s.getStatusCode() == 404
+                && (s.getErrorCode() == BlobErrorCode.RESOURCE_NOT_FOUND
+                || s.getErrorCode() == BlobErrorCode.BLOB_NOT_FOUND);
+            /* HttpResponseException - file get properties is a head request so a body is not returned. Error
+             conversion logic does not properly handle errors that don't return XML. */
+        } else if (t instanceof HttpResponseException) {
+            HttpResponseException h = (HttpResponseException) t;
+            String errorCode = h.getResponse().getHeaderValue(X_MS_ERROR_CODE);
+            return h.getResponse().getStatusCode() == 404
+                && (BlobErrorCode.RESOURCE_NOT_FOUND.toString().equals(errorCode)
+                || BlobErrorCode.BLOB_NOT_FOUND.toString().equals(errorCode));
+        } else {
+            return false;
+        }
+    }
+
+    public static boolean checkContainerDoesNotExistStatusCode(Throwable t) {
+        if (t instanceof BlobStorageException) {
+            BlobStorageException s = (BlobStorageException) t;
+            return s.getStatusCode() == 404
+                && (s.getErrorCode() == BlobErrorCode.RESOURCE_NOT_FOUND
+                || s.getErrorCode() == BlobErrorCode.CONTAINER_NOT_FOUND);
+            /* HttpResponseException - file get properties is a head request so a body is not returned. Error
+             conversion logic does not properly handle errors that don't return XML. */
+        } else if (t instanceof HttpResponseException) {
+            HttpResponseException h = (HttpResponseException) t;
+            String errorCode = h.getResponse().getHeaderValue(X_MS_ERROR_CODE);
+            return h.getResponse().getStatusCode() == 404
+                && (BlobErrorCode.RESOURCE_NOT_FOUND.toString().equals(errorCode)
+                || BlobErrorCode.CONTAINER_NOT_FOUND.toString().equals(errorCode));
+        } else {
+            return false;
+        }
+    }
+
+    public static String tagsToString(Map<String, String> tags) {
+        if (tags == null || tags.isEmpty()) {
+            return null;
+        }
+        StringBuilder sb = new StringBuilder();
+        for (Map.Entry<String, String> entry : tags.entrySet()) {
+            try {
+                sb.append(URLEncoder.encode(entry.getKey(), Charset.defaultCharset().toString()));
+                sb.append("=");
+                sb.append(URLEncoder.encode(entry.getValue(), Charset.defaultCharset().toString()));
+                sb.append("&");
```

Comment:
Instead of always appending `&` at the end, start by checking if the `StringBuilder` isn't empty and append `&` before adding the next tag. `StringBuilder.deleteCharAt` requires copying the entire backing array for the `StringBuilder` and if this API is called often this could introduce a performance bottleneck.